ethereum/contracts/relayer/coreRelayer/CoreRelayerDelivery.sol

```diff
@@ -1,623 +1,560 @@
-            }
-
-            deliveryInstruction.executionParameters.gasLimit = overrides.gasLimit;
-            deliveryInstruction.receiverValueTarget = overrides.receiverValue;
-            deliveryInstruction.maximumRefundTarget = overrides.maximumRefund;
-            return (deliveryInstruction, overrides.redeliveryHash);
-        }
+    //Refund the relayer (their extra funds) + (the amount that the relayer spent on gas)
+    //  + (the users refund if that refund didn't succeed)
+    uint256 relayerRefundAmount = (
+      msg.value - deliveryInstruction.receiverValueTarget - deliveryInstruction.maximumRefundTarget
+    ) + (deliveryInstruction.maximumRefundTarget - transactionFeeRefundAmount)
+    //TODO AMO: Isn't this a bug? We add the same amount regardless of whether we hit the max or not
+      + ((refundStatus == RefundStatus.REFUND_SENT ||
+          refundStatus == RefundStatus.CROSS_CHAIN_REFUND_SENT ||
+          refundStatus == RefundStatus.CROSS_CHAIN_REFUND_SENT_MAXIMUM_BUDGET
+         ) ? 0 : refundToRefundAddress);
+
+    //TODO AMO: what if pay fails? (i.e. returns false)
+    //Refund the relay provider
+    pay(relayerRefundAddress, relayerRefundAmount);
+  }
+
+  function payRefundToRefundAddress(
+    uint16 refundChainId,
+    bytes32 refundAddress,
+    uint256 refundAmount,
+    bytes32 relayerAddress
+  ) private returns (RefundStatus) {
+    //same chain refund
+    if (refundChainId == getChainId())
+      return pay(payable(fromWormholeFormat(refundAddress)), refundAmount)
+        ? RefundStatus.REFUND_SENT
+        : RefundStatus.REFUND_FAIL;
+    
+    //cross-chain refund
+    uint256 wormholeMessageFee = getWormhole().messageFee();
+    if (refundAmount <= wormholeMessageFee)
+      return RefundStatus.CROSS_CHAIN_REFUND_FAIL_NOT_ENOUGH;
+
+    IRelayProvider relayProvider = IRelayProvider(fromWormholeFormat(relayerAddress));
+    if (!relayProvider.isChainSupported(refundChainId))
+      return RefundStatus.CROSS_CHAIN_REFUND_FAIL_PROVIDER_NOT_SUPPORTED;
+
+    uint256 refundSubMessageFee;
+    unchecked{refundSubMessageFee = refundAmount - wormholeMessageFee;}
+
+    uint256 receiverValueTarget = calculateTargetDeliveryMaximumRefund(
+      refundChainId, refundSubMessageFee, relayProvider
+    );
+    
+    //If refundAmount is not enough to pay for one wei of receiver value, then do not perform the
+    //  cross-chain refund (i.e. if (delivery overhead) + (wormhole message fee) + (cost of one wei
+    //  of receiver value) is larger than the remaining refund)
+    //TODO AMO: but what happens to the value then?
+    if (receiverValueTarget == 0)
+      return RefundStatus.CROSS_CHAIN_REFUND_FAIL_NOT_ENOUGH;
+
+    uint256 maxBudget = relayProvider.quoteMaximumBudget(refundChainId);
+    bool exceedsMaxBudget = false;
+    if (receiverValueTarget > maxBudget) {
+      //TODO AMO: What happens with the difference if the maximum budget is exceeded?
```

Comment:
It gets paid to the relay provider as part of the cost (so the price doesn't change, the amount given just gets upper bounded by the maximum budget)